File: app/src/test/java/me/raatiniemi/worker/features/settings/project/presenter/DataPresenterTest.java
/*
 * Copyright (C) 2018 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.settings.project.presenter;

import org.greenrobot.eventbus.EventBus;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import java.io.File;

import me.raatiniemi.worker.features.settings.data.model.Backup;
import me.raatiniemi.worker.features.settings.data.model.BackupSuccessfulEvent;
import me.raatiniemi.worker.features.settings.data.presenter.DataPresenter;
import me.raatiniemi.worker.features.settings.data.view.DataView;

import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;

@RunWith(JUnit4.class)
public class DataPresenterTest {
    private DataPresenter presenter;
    private DataView view;
    private EventBus eventBus;

    @Before
    public void setUp() {
        eventBus = mock(EventBus.class);
        presenter = new DataPresenter(eventBus);
        view = mock(DataView.class);
    }

    @Test
    public void attachView_registerEventBus() {
        presenter.attachView(view);

        verify(eventBus).register(presenter);
    }

    @Test
    public void detachView_unregisterEventBus() {
        presenter.detachView();

        verify(eventBus).unregister(presenter);
    }

    @Test
    public void onEventMainThread_successfulBackupEvent() {
        Backup backup = new Backup(new File("backup-file"));
        BackupSuccessfulEvent event = new BackupSuccessfulEvent(backup);
        presenter.attachView(view);

        presenter.onEventMainThread(event);

        verify(view).setLatestBackup(backup);
    }

    @Test
    public void onEventMainThread_successfulBackupEventWithoutView() {
        Backup backup = new Backup(new File("backup-file"));
        BackupSuccessfulEvent event = new BackupSuccessfulEvent(backup);

        presenter.onEventMainThread(event);

        verify(view, never()).setLatestBackup(backup);
    }
}

File: app/src/test/java/me/raatiniemi/worker/features/settings/data/presenter/DataPresenterTest.kt
/*
 * Copyright (C) 2019 Tobias Raatiniemi
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, version 2 of the License.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package me.raatiniemi.worker.features.settings.data.presenter

import me.raatiniemi.worker.features.settings.data.model.Backup
import me.raatiniemi.worker.features.settings.data.model.BackupSuccessfulEvent
import me.raatiniemi.worker.features.settings.data.view.DataView
import org.greenrobot.eventbus.EventBus
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.JUnit4
import org.mockito.Mockito.*
import java.io.File

@RunWith(JUnit4::class)
class DataPresenterTest {
    private lateinit var eventBus: EventBus
    private lateinit var view: DataView
    private lateinit var presenter: DataPresenter

    @Before
    fun setUp() {
        eventBus = mock(EventBus::class.java)
        view = mock(DataView::class.java)
        presenter = DataPresenter(eventBus)
    }

    @Test
    fun attachView_registerEventBus() {
        presenter.attachView(view)

        verify<EventBus>(eventBus).register(presenter)
    }

    @Test
    fun detachView_unregisterEventBus() {
        presenter.detachView()

        verify<EventBus>(eventBus).unregister(presenter)
    }

    @Test
    fun onEventMainThread_successfulBackupEvent() {
        val backup = Backup(File("backup-file"))
        val event = BackupSuccessfulEvent(backup)
        presenter.attachView(view)

        presenter.onEventMainThread(event)

        verify<DataView>(view).setLatestBackup(backup)
    }

    @Test
    fun onEventMainThread_successfulBackupEventWithoutView() {
        val backup = Backup(File("backup-file"))
        val event = BackupSuccessfulEvent(backup)

        presenter.onEventMainThread(event)

        verify<DataView>(view, never()).setLatestBackup(backup)
    }
}
